syntax = "proto3";

package gfs;

// Messages
message StringMessage {
    string value = 1;
}

message FileRequest {
    string filename = 1;
}

message FileResponse {
    bool success = 1;
    string message = 2;
}

message ChunkRequest {
    string chunk_id = 1;
}

message AppendRequest {
    string file_name = 1;
    string content = 2;
    string secondary_chunk = 3;
}

message ChunkResponse {
    bool success = 1;
    string message = 2;
    int64 available_space = 3;
}


message ReadRequest {
    string filename = 1;
    int64 offset = 2;
    int32 length = 3;
}

message ReadResponse {
    bool success = 1;
    bytes data = 2;
}

// Master Server Interface
service MasterServerToClient {
    rpc ListFiles(StringMessage) returns (StringMessage);
    rpc CreateFile(FileRequest) returns (FileResponse);
    rpc RequestLease(FileRequest) returns (FileResponse);
    rpc VerifyLease(FileRequest) returns (FileResponse);
    rpc AppendFile(FileRequest) returns (FileResponse);
    rpc AppendAck(FileRequest) returns (FileResponse);
    rpc CreateChunk(ChunkRequest) returns (ChunkResponse);
    rpc ReadFile(ReadRequest) returns (ReadResponse);
    rpc DeleteFile(FileRequest) returns (FileResponse);
    rpc UndeleteFile(FileRequest) returns (FileResponse);
}

// Chunk Server Interface
service ChunkServerToClient {
    rpc Create(ChunkRequest) returns (ChunkResponse);
    rpc Append(AppendRequest) returns (ChunkResponse);
    rpc Read(ReadRequest) returns (ReadResponse);
}

// Master Server to Chunk Server communication
service MasterServerToChunkServer {
    
    rpc DuplicateFile(FileRequest) returns (ChunkResponse);
    rpc DeleteFile(ChunkRequest) returns (ChunkResponse);   // e.g., for deleting a chunk
}

// Chunk Server to Chunk Server communication
service ChunkServerToChunkServer {
    rpc Append_ChunkToChunk(AppendRequest) returns (ChunkResponse);  // For chunk replication across servers
    rpc SyncChunkData(ChunkRequest) returns (ChunkResponse);   // Synchronize data between chunk servers
}

